### ISSUE TITLE
Multiple uses of existential/question-mark variables in function definitions do not work as expected 

### ISSUE THREAD

**Issue by nishantjr** [kframework/kore#677]
Minimal example below:

```
$ cat foo.k
module FOO
  imports INT
  imports BOOL
  
  syntax Seq ::= List{Int, ":"}
  syntax Seq ::= "foo" "(" Int ")" [function]
  rule foo(N) => ?_:Int : foo(N -Int 1)
    requires N >Int 0
  rule foo(0) => .Seq

  syntax X ::= "foo"
  rule <k> foo => foo(3) </k>
endmodule

$ ./ext/k/k-distribution/target/release/k/bin/kompile foo.k --backend haskell
[Warning] Compiler: Could not find main syntax module with name FOO-SYNTAX in
definition.  Use --syntax-module to specify one. Using FOO as default.

$ cat foo
foo

$ ./ext/k/k-distribution/target/release/k/bin/krun foo
Reading the input file TimeSpec {sec = 0, nsec = 84834}
Parsing the file TimeSpec {sec = 0, nsec = 45296501}
Verifying the definition TimeSpec {sec = 0, nsec = 25025880}
Reading the input file TimeSpec {sec = 0, nsec = 31449}
Parsing the file TimeSpec {sec = 0, nsec = 121358}
Verifying the pattern TimeSpec {sec = 0, nsec = 126137}
Executing TimeSpec {sec = 0, nsec = 27401608}
<k>
  'Unds'12 : 'Unds'12 : 'Unds'12 : .Seq
</k>
```

Expected output is 

```
<k>
  'Unds'12 : 'Unds'13 : 'Unds'14 : .Seq
</k>
```

**Comment by virgil-serbanuta** [comment]
Summary: at first sight, it seems to be both a frontend issue (universal quantification instead of existential) and a backend issue (i.e. not renaming conflicting variables). I think that we implicitly assumed existential quantification for new variables in rules (assumption that should go away at some point, I guess), so the main problem should be in the backend.

The `rule foo(N) => ?_:Int : foo(N -Int 1)` axiom gets translated to
```
axiom{R} \implies{R} (
    \equals{SortBool{},R}(
        Lbl'Unds-GT-'Int'UndsUnds'INT-COMMON'UndsUnds'Int'Unds'Int{}(
            VarN:SortInt{},
            \dv{SortInt{}}("0")
        ),
        \dv{SortBool{}}("true")),
    \and{R} (
        \equals{SortSeq{},R} (
            Lblfoo'LParUndsRParUnds'FRESH-VARS'UndsUnds'Int{}(VarN:SortInt{}),
            Lbl'UndsColnUndsUnds'FRESH-VARS'UndsUnds'Int'Unds'Seq{}(
                Var'Unds'12:SortInt{},
                Lblfoo'LParUndsRParUnds'FRESH-VARS'UndsUnds'Int{}(
                    Lbl'Unds'-Int'UndsUnds'INT-COMMON'UndsUnds'Int'Unds'Int{}(
                        VarN:SortInt{},
                        \dv{SortInt{}}("1")
                    )
                )
            )
        ),
        \top{R}()
    )
)
```
or, in a more readable form:
```
axiom{R} \implies{R} (
    ((varN <Int 0) == true)
    \equals{SortSeq{},R} (
        foo(VarN),
        Coln(Var_12:SortInt{}, foo(VarN -Int 1))
    )
)
```
Note that `Var_12:SortInt{}` is universally quantified, and `foo(VarN)` is supposed to be a singleton, so the above axiom should not work (isn't the axiom pattern always bottom if Coln is not constant in the first argument?). @traiansf in case he has comments, but, I guess, it should be quantified existentially.

Regardless, if I were to apply this to `foo(3)`, the first step would go like this:
```
foo(3) = Coln(Var_12, foo(2))
```

But then, when replacing `foo(2)` with `Coln(Var_12, foo(1))`, should't Var_12 be renamed? I mean, that was one of the important thing that `Target` variables were supposed to do, so I'll assume that the backend has a bug here, too. @ttuegel in case he has comments about this.

Regardless of the above comments, I'll try to get a better understanding of the backend issue.

**Comment by ttuegel** [comment]
@virgil-serbanuta Ah, I realized what the problem is! One of the first things we do when we apply a rule is rename all the rule's implicitly-quantified variables so they do not conflict with any variables in the configuration. But, when we apply a function rule, we only apply the rule to a sub-term, so we cannot see all the variables in the pattern! One solution would be: any time we introduce a variable when applying a rule, that variable should be existentially quantified in the resulting term. We never did this before, so I think this bug has always been lurking. Existentially quantifying all the new variables might cause problems for us because we can't always simplify the binders away. In principle, if we could lift the quantifier all the way to the top of the configuration, we could remove it.

**Comment by virgil-serbanuta** [comment]
I agree with this in principle, but we should be a bit careful: existential quantifiers could become universal when lifting in predicates (in theory, they could become universal in terms, too), we can't change the quantifier order, and we can only remove existential ones at the top. So, in cases where lifting would produce an universal quantifier at the top, it might be better to leave the existential where it occurred.

However, the Scala backend lifted all quantifiers, so I wonder if there's something I'm missing. Anyway, I would only lift existential quantifiers if they have an even number of negation above, and the universal ones when they have an odd number of negations.

Indeed, the bug has been lurking for a long time. However, the initial implementation for variable uniqueness attempted to generate globally unique IDs, which, for all its drawbacks, didn't have this problem. On the other hand, we never used explicit quantifiers for variables introduced by rule application, so, implicitly, we also assumed that all functions (including the ones in predicates) had an even number of negations above them.

**Comment by ttuegel** [comment]
I wouldn't want to just lift the quantifiers immediately, but have the simplifier move them up level-by-level, until they can be removed (if desired) at the top level. Does that address your concern about the parity of negation?

**Comment by virgil-serbanuta** [comment]
I meant this: if we have a term `sigma(exists x . phi, not (exists y . psi))`, we can get any of these when lifting quantifiers:
```
forall y . exists x . sigma(phi, not psi)
exists x . forall y . sigma(phi, not psi)
exists x . sigma(phi, not (exists y . psi))
```
The first formula seems useless: We can't remove any quantifier present now, and none of the new quantifiers that may show up in the formula.
The second formula seems slightly more useful: we can remove `exists x`, but it will be the last quantifier that we'll ever remove. With some effort, we could make sure that our simplification generates this instead of the first formula.
The third formula seems the most useful, since we can remove `exists x` and many of the future quantifiers that may show up in the formula. The effort needed to implement this seems similar to the effort needed to distinguish the second formula from the first, but I might be wrong.

So then we have the following options (maybe others, too):
1. Stop executing and crash when we get a `forall` at the top.
2. Figure out how to deal with forall at the top (perhaps we just need to learn to ignore quantifiers at the top and hope that either we don't see forall at the top or the SMT can handle it).
3. Never lift quantifiers that will become `forall` at the top (again, hoping that the SMT can either handle or ignore it). Perhaps we can even try to narrow the scope for these quantifiers.

I think that all of the above could work. The first one will let us know if this can happen in real-life cases and we'll have a concrete example that may provide hints about what to do. The latter will let us ignore the quantifier for a long time, hoping that the SMT can deal with it (assuming that we have negations only in predicates, which seems to be reasonable for now).

**Comment by ttuegel** [comment]
> Never lift quantifiers that will become `forall` at the top (again, hoping that the SMT can either handle or ignore it).

Right now, I don't think we lift quantifiers around `\not`, so this should not be a problem. I'm assuming we prefer to work with `\exists` whenever possible. We could add simplification cases that lift `\not \forall` into `\exists \not` and lower `\forall \not` into `\not exists`.

**Comment by virgil-serbanuta** [comment]
Well, IIRC, right now we don't lift any quantifiers at all...

**Comment by virgil-serbanuta** [comment]
I just merged #684 which fixes the "the resulting pattern is broken" issue. Should we keep this issue open until we lift quantifiers, or should we close it now?

**Comment by ttuegel** [comment]
Let's close it now. We will discover more and different problems that will require lifting quantifiers.
